### ISSUE TITLE
Support flat config

### ISSUE THREAD

**Issue by cprussin** [jest-community/jest-runner-eslint#166]
As you all probably already know, [there is a new eslint config format in the works](https://eslint.org/blog/2022/08/new-config-system-part-1/).

It seems like in order for libraries to work with it, the libraries [must use the `FlatESLint` class instead of the `ESLint` class](https://eslint.org/blog/2022/08/new-config-system-part-3/#using-flat-config-with-the-eslint-class), at least tentatively while the new config system is being stabilized and everything is getting migrated.

Would there be any interest in supporting the new flat config format during the transitional phase?  Presumably it would be a pretty simple change -- just would need to do something like add a setting that switches between `ESLint` and `FlatESLint` [here](https://github.com/jest-community/jest-runner-eslint/blob/main/src/runner/runESLint.js#L119).  But I could also see an argument against supporting an experimental API at all, especially given the article above mentions that the `FlatESLint` class would only exist during the transition phase and would go away once flat config becomes the default.

I'd be happy to submit a PR if the maintainers here would be amenable to accepting it.

@SimenB I'll tag you to get your opinion on this since you helped me merge in my last contributions and I know you know the ecosystem really well, but I know you're pretty busy so please help me tag in someone else who can give guidance / opinions if there's someone more appropriate!

Thanks as always

**EDIT:** I hacked up a super quick implementation of what a switch to `FlatESLint` would look like [here](https://github.com/jest-community/jest-runner-eslint/compare/main...cprussin:jest-runner-eslint:main) which I've tested in my own configs, if you want to take a look.  This implementation doesn't add an option, it just switches over, and it doesn't do anything to try to translate changed cli options to proper equivalent (I'm just deleting those options entirely).  So this wouldn't be ready to be put into a PR, but it gives you the gist of the changes that would need to happen.

**Comment by ljharb** [comment]
Supporting both seems ideal.

**Comment by cprussin** [comment]
> Supporting both seems ideal.

Right, sorry if I wasn't clear but that's what I'm proposing -- basically just add an option to jest-runner-eslint that would allow selecting which config format a user wants.

**Comment by ljharb** [comment]
Is there a reason we can't detect it on our own?

**Comment by cprussin** [comment]
That's a good question, I suppose we could just look for presence of a `eslint.config.js` which is exclusively for use with the flat config.  I really only suggested an option since the whole feature in eslint is still unstable and opt-in and I figured jest-runner-eslint should explicitly opt in as well, but I suppose it doesn't really buy anything to have to opt in to the same experimental feature twice does it?

tl;dr: yeah I think we could detect it

**Comment by ljharb** [comment]
We'd need an option anyways since `eslint.config.js` could still be used for a legacy config, but as a default that seems reasonable.

**Comment by cprussin** [comment]
@ljharb are you sure it could be used for legacy config?  It's not documented anywhere in https://eslint.org/docs/latest/use/configure/configuration-files#configuration-file-formats and I've never seen an eslint config using that file.  I always just assumed that eslint just used cosmiconfig but based on the way the config files are documented, I also wouldn't be surprised if it were using something else and this is the first that `eslint.config.js` is actually a respected config file.  I don't know for sure and I don't see anything definitive in the docs.

**Comment by ljharb** [comment]
With `-c` you can specify any file you like.

**Comment by cprussin** [comment]
ah, fair enough!

I think that eslint would assume `eslint.config.js` is the flat format unless told otherwise, so I'd propose we do the same.  In other words, the logic would be to use `FlatESLint` if 1. `eslint.config.js` exists; and 2. `config` isn't specified in `jest-runner-eslint` options.  Does that make sense to you?

**Comment by ljharb** [comment]
Yep, that sounds right to me.

**Comment by cprussin** [comment]
@ljharb Any advice on what to do with the options that seem to not work with `FlatESLint`?  The list of options that I found to be problematic are:

- extensions
- ignorePath
- rulePaths
- resolvePluginsRelativeTo
- useEslintrc
- overrideConfig

I'm not really familiar with most of these -- in the PoC diff I posted above I just deleted them from `cliOptions`.  I'm assuming `resolvePluginsRelativeTo` would no longer be needed since the flat config format uses ESM imports and doesn't suffer from dynamic plugin resolution issues.  I'm not as sure about the others.  Do you think they're safe to just omit if we're using flat config?

**Comment by SimenB** [comment]
`overrideConfig` superficially sounds useful, but jest runner specific options doesn't sound like a good idea.

Let's warn on their presence saying they're ignored when using flat/modern config?

**Comment by ljharb** [comment]
I'm confused; why would the flat config not have a way to specify most of those things?

**Comment by cprussin** [comment]
> I'm confused; why would the flat config not have a way to specify most of those things?

Not sure on that @ljharb , I just got errors about those keys and as I was just whipping up a quick POC I didn't bother with investigating.  I can look more into it when I prep a true PR.  Presumably some of these keys changed name or format or something with the flat config.
